daemons/controld/controld_based.c

```diff
@@ -92,75 +97,83 @@ do_cib_control(long long action,
                enum crmd_fsa_state cur_state,
                enum crmd_fsa_input current_input, fsa_data_t * msg_data)
 {
-    CRM_ASSERT(fsa_cib_conn != NULL);
+    cib_t *cib_conn = controld_globals.cib_conn;
+
+    void (*dnotify_fn) (gpointer user_data) = crmd_cib_connection_destroy;
+    void (*replace_cb) (const char *event, xmlNodePtr msg) = do_cib_replaced;
+    void (*update_cb) (const char *event, xmlNodePtr msg) = do_cib_updated;
 
-    if (action & A_CIB_STOP) {
+    int rc = pcmk_ok;
 
-        if (fsa_cib_conn->state != cib_disconnected && last_resource_update != 0) {
-            crm_info("Waiting for resource update %d to complete", last_resource_update);
+    CRM_ASSERT(cib_conn != NULL);
+
+    if (pcmk_is_set(action, A_CIB_STOP)) {
+        if ((cib_conn->state != cib_disconnected)
+            && (last_resource_update != 0)) {
+
+            crm_info("Waiting for resource update %d to complete",
+                     last_resource_update);
             crmd_fsa_stall(FALSE);
             return;
         }
-
         controld_disconnect_cib_manager();
-
     }
 
-    if (action & A_CIB_START) {
-        int rc = pcmk_ok;
-
-        if (cur_state == S_STOPPING) {
-            crm_err("Ignoring request to connect to the CIB manager after shutdown");
-            return;
-        }
+    if (!pcmk_is_set(action, A_CIB_START)) {
+        return;
+    }
 
-        rc = fsa_cib_conn->cmds->signon(fsa_cib_conn, CRM_SYSTEM_CRMD, cib_command_nonblocking);
+    if (cur_state == S_STOPPING) {
+        crm_err("Ignoring request to connect to the CIB manager after "
+                "shutdown");
+        return;
+    }
 
-        if (rc != pcmk_ok) {
-            /* a short wait that usually avoids stalling the FSA */
-            sleep(1);
-            rc = fsa_cib_conn->cmds->signon(fsa_cib_conn, CRM_SYSTEM_CRMD, cib_command_nonblocking);
-        }
+    rc = cib_conn->cmds->signon(cib_conn, CRM_SYSTEM_CRMD,
+                                cib_command_nonblocking);
```

Comment:
The signon() method doesn't actually check whether we're already signed on, and this function doesn't check until the very end when we check `R_CIB_CONNECTED`. May want to follow up on that in the future.

Reply:
We ensure we don't get here unless we know it's disconnected, but it never hurts to have a failsafe